src/MinFraud.php

```diff
@@ -198,59 +482,631 @@ public function withEmail(array $values): self
-    {
-        return $this->validateAndAdd('Billing', 'billing', $values);
+    public function withBilling(
+        array $values = [],
+        ?string $address = null,
+        ?string $address2 = null,
+        ?string $city = null,
+        ?string $company = null,
+        ?string $country = null,
+        ?string $firstName = null,
+        ?string $lastName = null,
+        ?string $phoneCountryCode = null,
+        ?string $phoneNumber = null,
+        ?string $postal = null,
+        ?string $region = null,
+    ): self {
+        if (\count($values) !== 0) {
+            if (\func_num_args() !== 1) {
+                throw new \InvalidArgumentException(
+                    'You may only provide the $values array or named arguments, not both.',
+                );
+            }
+
+            $address = $this->remove($values, 'address');
+            $address2 = $this->remove($values, 'address_2');
+            $city = $this->remove($values, 'city');
+            $company = $this->remove($values, 'company');
+            $country = $this->remove($values, 'country');
+            $firstName = $this->remove($values, 'first_name');
+            $lastName = $this->remove($values, 'last_name');
+            $phoneCountryCode = $this->remove($values, 'phone_country_code');
+            $phoneNumber = $this->remove($values, 'phone_number');
+            $postal = $this->remove($values, 'postal');
+            $region = $this->remove($values, 'region');
+
+            $this->verifyEmpty($values);
+        }
+
+        if ($address !== null) {
+            $values['address'] = $address;
+        }
+
+        if ($address2 !== null) {
+            $values['address_2'] = $address2;
+        }
+
+        if ($city !== null) {
+            $values['city'] = $city;
+        }
+
+        if ($company !== null) {
+            $values['company'] = $company;
+        }
+
+        if ($country !== null) {
+            if (!preg_match('/^[A-Z]{2}$/', $country)) {
+                $this->maybeThrowInvalidInputException("$country is not a valid ISO 3166-1 country code");
+            }
+            $values['country'] = $country;
+        }
+
+        if ($firstName !== null) {
+            $values['first_name'] = $firstName;
+        }
+
+        if ($lastName !== null) {
+            $values['last_name'] = $lastName;
+        }
+
+        if ($phoneCountryCode !== null) {
+            if (!preg_match('/^[0-9]{1,4}$/', $phoneCountryCode)) {
+                $this->maybeThrowInvalidInputException('Phone country code must be a string of 1 to 4 digits.');
+            }
+            $values['phone_country_code'] = $phoneCountryCode;
+        }
+
+        if ($phoneNumber !== null) {
+            $values['phone_number'] = $phoneNumber;
+        }
+
+        if ($postal !== null) {
+            $values['postal'] = $postal;
+        }
+
+        if ($region !== null) {
+            if (!preg_match('/^[0-9A-Z]{1,4}$/', $region)) {
```

Comment:
I see `{1,4}` matches the original validation rule (and what [the docs](https://dev.maxmind.com/minfraud/api-documentation/requests/#schema--request--billing__region) say is the max length), but shouldn't this be `{1,3}`? Similarly to the other validation on `$region` in `withShipping`. Although I also see (and the max length mentioned by [the docs

Reply:
Yeah, I was wondering about the 4 in the existing validation too but didn't dig into it. I'll double check what the web service does.

Reply:
It appears the web service uses `\A[0-9A-Z]{1,4}\z`. That is probably something worth looking into, but I am inclined to leave it for now.